synapse/storage/util/id_generators.py

```diff
@@ -352,12 +356,39 @@ def _mark_id_as_finished(self, next_id: int):
 
         with self._lock:
             self._unfinished_ids.discard(next_id)
+            self._finished_ids.add(next_id)
+
+            new_cur = None
+
+            if self._unfinished_ids:
+                # If there are unfinished IDs then the new position will be the
+                # largest finished ID less than the minimum unfinished ID.
+
+                finished = set()
+
+                min_unfinshed = min(self._unfinished_ids)
+                for s in self._finished_ids:
+                    if s < min_unfinshed:
+                        if not new_cur or new_cur < s:
+                            new_cur = s
+                    else:
+                        finished.add(s)
+
+                # We clear these out since they're now all less than the new
+                # position.
+                self._finished_ids = finished
```

Comment:
Is this essentially discarding all finished IDs that are less than the minimum unfinished ID? I'm guessing we can't discard in a loop (which would be clearer to me) because Python doesn't like you modifying things as you iterate them?

Reply:
This all feels like it is doing a lot more work than before, but I guess both the old code and new are essentially iterating `_unfinished_ids` once (although this also builds a second set).

Reply:
Yeah, basically we want to remove all items that are less than minimum unfinished ID. We could rewrite it to be a sorted list instead and then chop off the front of it, but its not obvious that that is better/quicker/easier.